# Import & Define Tables

In [1]:
import kagglehub
import pandas as pd
import os


# Download latest version
path = kagglehub.dataset_download("nadyinky/sephora-products-and-skincare-reviews")

print("Path to dataset files:", path)


Path to dataset files: C:\Users\nicol\.cache\kagglehub\datasets\nadyinky\sephora-products-and-skincare-reviews\versions\2


In [2]:
# assign path
path, dirs, files = next(os.walk(path))
file_count = len(files)
dataframes_list = []

print(path)
print(files)


# turn product.csv into a dataframe
product_df = pd.read_csv(path + '\\' + 'product_info.csv')


# merge reviews into one csv file
all_df = []


reviews = files[1:]

for file in reviews:
    full_path = os.path.join(path,file)
    df = pd.read_csv(full_path)
    all_df.append(df)

reviews_df = pd.concat(all_df, ignore_index=True)

reviews_df.to_csv(os.path.join(path, "merged_file.csv"), index=False)

C:\Users\nicol\.cache\kagglehub\datasets\nadyinky\sephora-products-and-skincare-reviews\versions\2
['merged_file.csv', 'product_info.csv', 'reviews_0-250.csv', 'reviews_1250-end.csv', 'reviews_250-500.csv', 'reviews_500-750.csv', 'reviews_750-1250.csv']


C:\Users\nicol\AppData\Local\Temp\ipykernel_11524\2066281815.py:22: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(full_path)
C:\Users\nicol\AppData\Local\Temp\ipykernel_11524\2066281815.py:22: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(full_path)
C:\Users\nicol\AppData\Local\Temp\ipykernel_11524\2066281815.py:22: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(full_path)


# Imported Tables


## Products

In [3]:
product_df.head()

product_id               product_name  brand_id brand_name  loves_count  \
0    P473671    Fragrance Discovery Set      6342      19-69         6320   
1    P473668    La Habana Eau de Parfum      6342      19-69         3827   
2    P473662  Rainbow Bar Eau de Parfum      6342      19-69         3253   
3    P473660       Kasbah Eau de Parfum      6342      19-69         3018   
4    P473658  Purple Haze Eau de Parfum      6342      19-69         2691   

   rating  reviews            size                      variation_type  \
0  3.6364     11.0             NaN                                 NaN   
1  4.1538     13.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   
2  4.2500     16.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   
3  4.4762     21.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   
4  3.2308     13.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   

  variation_value  ... online_only out_of_stock  sephora_exclusive  \
0             NaN  ...           1            0                  0   
1  3.4 oz/ 100 mL  ...           1            0                  0   
2  3.4 oz/ 100 mL  ...           1            0                  0   
3  3.4 oz/ 100 mL  ...           1            0                  0   
4  3.4 oz/ 100 mL  ...           1            0                  0   

                                          highlights  primary_category  \
0  ['Unisex/ Genderless Scent', 'Warm &Spicy Scen...         Fragrance   
1  ['Unisex/ Genderless Scent', 'Layerable Scent'...         Fragrance   
2  ['Unisex/ Genderless Scent', 'Layerable Scent'...         Fragrance   
3  ['Unisex/ Genderless Scent', 'Layerable Scent'...         Fragrance   
4  ['Unisex/ Genderless Scent', 'Layerable Scent'...         Fragrance   

   secondary_category  tertiary_category  child_count  child_max_price  \
0   Value & Gift Sets  Perfume Gift Sets            0              NaN   
1               Women            Perfume            2             85.0   
2               Women            Perfume            2             75.0   
3               Women            Perfume            2             75.0   
4               Women            Perfume            2             75.0   

   child_min_price  
0              NaN  
1             30.0  
2             30.0  
3             30.0  
4             30.0  

[5 rows x 27 columns]

## Reviews

In [4]:
reviews_df.head()

product_id               product_name  brand_id brand_name  loves_count  \
0    P473671    Fragrance Discovery Set    6342.0      19-69       6320.0   
1    P473668    La Habana Eau de Parfum    6342.0      19-69       3827.0   
2    P473662  Rainbow Bar Eau de Parfum    6342.0      19-69       3253.0   
3    P473660       Kasbah Eau de Parfum    6342.0      19-69       3018.0   
4    P473658  Purple Haze Eau de Parfum    6342.0      19-69       2691.0   

   rating  reviews            size                      variation_type  \
0  3.6364     11.0             NaN                                 NaN   
1  4.1538     13.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   
2  4.2500     16.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   
3  4.4762     21.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   
4  3.2308     13.0  3.4 oz/ 100 mL  Size + Concentration + Formulation   

  variation_value  ... total_feedback_count total_neg_feedback_count  \
0             NaN  ...                  NaN                      NaN   
1  3.4 oz/ 100 mL  ...                  NaN                      NaN   
2  3.4 oz/ 100 mL  ...                  NaN                      NaN   
3  3.4 oz/ 100 mL  ...                  NaN                      NaN   
4  3.4 oz/ 100 mL  ...                  NaN                      NaN   

   total_pos_feedback_count  submission_time  review_text  review_title  \
0                       NaN              NaN          NaN           NaN   
1                       NaN              NaN          NaN           NaN   
2                       NaN              NaN          NaN           NaN   
3                       NaN              NaN          NaN           NaN   
4                       NaN              NaN          NaN           NaN   

   skin_tone  eye_color  skin_type  hair_color  
0        NaN        NaN        NaN         NaN  
1        NaN        NaN        NaN         NaN  
2        NaN        NaN        NaN         NaN  
3        NaN        NaN        NaN         NaN  
4        NaN        NaN        NaN         NaN  

[5 rows x 41 columns]

# Data Description

## Products

### Descriptive Statistics Analysis

In [5]:
#Describe the data

product_df.describe()

brand_id   loves_count       rating       reviews    price_usd  \
count  8494.000000  8.494000e+03  8216.000000   8216.000000  8494.000000   
mean   5422.440546  2.917957e+04     4.194513    448.545521    51.655595   
std    1709.595957  6.609212e+04     0.516694   1101.982529    53.669234   
min    1063.000000  0.000000e+00     1.000000      1.000000     3.000000   
25%    5333.000000  3.758000e+03     3.981725     26.000000    25.000000   
50%    6157.500000  9.880000e+03     4.289350    122.000000    35.000000   
75%    6328.000000  2.684125e+04     4.530525    418.000000    58.000000   
max    8020.000000  1.401068e+06     5.000000  21281.000000  1900.000000   

       value_price_usd  sale_price_usd  limited_edition          new  \
count       451.000000      270.000000      8494.000000  8494.000000   
mean         91.168537       20.207889         0.070285     0.071698   
std          79.195631       24.327352         0.255642     0.258002   
min           0.000000        1.750000         0.000000     0.000000   
25%          45.000000        8.000000         0.000000     0.000000   
50%          67.000000       14.000000         0.000000     0.000000   
75%         108.500000       25.000000         0.000000     0.000000   
max         617.000000      320.000000         1.000000     1.000000   

       online_only  out_of_stock  sephora_exclusive  child_count  \
count  8494.000000   8494.000000        8494.000000  8494.000000   
mean      0.219096      0.073699           0.279374     1.631622   
std       0.413658      0.261296           0.448718     5.379470   
min       0.000000      0.000000           0.000000     0.000000   
25%       0.000000      0.000000           0.000000     0.000000   
50%       0.000000      0.000000           0.000000     0.000000   
75%       0.000000      0.000000           1.000000     1.000000   
max       1.000000      1.000000           1.000000   105.000000   

       child_max_price  child_min_price  
count      2754.000000      2754.000000  
mean         53.792023        39.665802  
std          58.765894        38.685720  
min           3.000000         3.000000  
25%          22.000000        19.000000  
50%          32.000000        28.000000  
75%          59.000000        42.000000  
max         570.000000       400.000000

### Columns

In [6]:
product_df.columns

Index(['product_id', 'product_name', 'brand_id', 'brand_name', 'loves_count',
       'rating', 'reviews', 'size', 'variation_type', 'variation_value',
       'variation_desc', 'ingredients', 'price_usd', 'value_price_usd',
       'sale_price_usd', 'limited_edition', 'new', 'online_only',
       'out_of_stock', 'sephora_exclusive', 'highlights', 'primary_category',
       'secondary_category', 'tertiary_category', 'child_count',
       'child_max_price', 'child_min_price'],
      dtype='object')

### Categories

In [7]:
product_df.value_counts('primary_category', normalize=True)

primary_category
Skincare           0.284907
Makeup             0.278903
Hair               0.172357
Fragrance          0.168590
Bath & Body        0.047681
Mini Size          0.033906
Men                0.007064
Tools & Brushes    0.006122
Gifts              0.000471
Name: proportion, dtype: float64

In [8]:
product_df.value_counts('secondary_category', normalize = True)

secondary_category
Women                        0.103111
Hair Styling & Treatments    0.089206
Eye                          0.083785
Face                         0.077657
Moisturizers                 0.064930
Value & Gift Sets            0.058685
Treatments                   0.054914
Shampoo & Conditioner        0.050790
Lip                          0.048433
Cleansers                    0.042541
Candles & Home Scents        0.030992
Brushes & Applicators        0.028989
Body Moisturizers            0.025925
Mini Size                    0.022036
Eye Care                     0.021918
Masks                        0.019562
Cheek                        0.019444
Tools                        0.018030
Makeup                       0.016144
Men                          0.015909
Sunscreen                    0.012845
Skincare                     0.011548
Bath & Shower                0.009899
High Tech Tools              0.009427
Wellness                     0.009309
Body Care                    0.

In [9]:
product_df.value_counts('tertiary_category', normalize = True)

tertiary_category
Perfume                      0.075693
Moisturizers                 0.051439
Face Serums                  0.050506
Rollerballs & Travel Size    0.038246
Hair Styling Products        0.033982
                               ...   
Under-Eye Concealer          0.000400
Sunscreen                    0.000267
Hair Thinning & Hair Loss    0.000267
Manicure & Pedicure Tools    0.000133
Damaged Hair                 0.000133
Name: proportion, Length: 118, dtype: float64

### Count Nulls

In [10]:
(product_df[:].isnull().sum() / len(product_df)).map('{:.2%}'.format)

product_id             0.00%
product_name           0.00%
brand_id               0.00%
brand_name             0.00%
loves_count            0.00%
rating                 3.27%
reviews                3.27%
size                  19.20%
variation_type        17.00%
variation_value       18.81%
variation_desc        85.28%
ingredients           11.13%
price_usd              0.00%
value_price_usd       94.69%
sale_price_usd        96.82%
limited_edition        0.00%
new                    0.00%
online_only            0.00%
out_of_stock           0.00%
sephora_exclusive      0.00%
highlights            25.98%
primary_category       0.00%
secondary_category     0.09%
tertiary_category     11.66%
child_count            0.00%
child_max_price       67.58%
child_min_price       67.58%
dtype: object

### Data Types

In [11]:
product_df.dtypes

product_id             object
product_name           object
brand_id                int64
brand_name             object
loves_count             int64
rating                float64
reviews               float64
size                   object
variation_type         object
variation_value        object
variation_desc         object
ingredients            object
price_usd             float64
value_price_usd       float64
sale_price_usd        float64
limited_edition         int64
new                     int64
online_only             int64
out_of_stock            int64
sephora_exclusive       int64
highlights             object
primary_category       object
secondary_category     object
tertiary_category      object
child_count             int64
child_max_price       float64
child_min_price       float64
dtype: object

## Reviews

### Descriptive Statistics Analysis

In [12]:
# Describe the data

reviews_df.describe()


brand_id   loves_count        rating       reviews     price_usd  \
count  8494.000000  8.494000e+03  1.102627e+06   8216.000000  1.102905e+06   
mean   5422.440546  2.917957e+04  4.298378e+00    448.545521  4.902877e+01   
std    1709.595957  6.609212e+04  1.146057e+00   1101.982529  4.016662e+01   
min    1063.000000  0.000000e+00  1.000000e+00      1.000000  3.000000e+00   
25%    5333.000000  3.758000e+03  4.000000e+00     26.000000  2.500000e+01   
50%    6157.500000  9.880000e+03  5.000000e+00    122.000000  3.900000e+01   
75%    6328.000000  2.684125e+04  5.000000e+00    418.000000  6.200000e+01   
max    8020.000000  1.401068e+06  5.000000e+00  21281.000000  1.900000e+03   

       value_price_usd  sale_price_usd  limited_edition          new  \
count       451.000000      270.000000      8494.000000  8494.000000   
mean         91.168537       20.207889         0.070285     0.071698   
std          79.195631       24.327352         0.255642     0.258002   
min           0.000000        1.750000         0.000000     0.000000   
25%          45.000000        8.000000         0.000000     0.000000   
50%          67.000000       14.000000         0.000000     0.000000   
75%         108.500000       25.000000         0.000000     0.000000   
max         617.000000      320.000000         1.000000     1.000000   

       online_only  ...  sephora_exclusive  child_count  child_max_price  \
count  8494.000000  ...        8494.000000  8494.000000      2754.000000   
mean      0.219096  ...           0.279374     1.631622        53.792023   
std       0.413658  ...           0.448718     5.379470        58.765894   
min       0.000000  ...           0.000000     0.000000         3.000000   
25%       0.000000  ...           0.000000     0.000000        22.000000   
50%       0.000000  ...           0.000000     0.000000        32.000000   
75%       0.000000  ...           1.000000     1.000000        59.000000   
max       1.000000  ...           1.000000   105.000000       570.000000   

       child_min_price    Unnamed: 0  is_recommended    helpfulness  \
count      2754.000000  1.094411e+06   926423.000000  532819.000000   
mean         39.665802  1.989864e+05        0.839962       0.767782   
std          38.685720  1.749886e+05        0.366642       0.317164   
min           3.000000  0.000000e+00        0.000000       0.000000   
25%          19.000000  5.590600e+04        1.000000       0.652174   
50%          28.000000  1.308240e+05        1.000000       0.928571   
75%          42.000000  3.285265e+05        1.000000       1.000000   
max         400.000000  6.021290e+05        1.000000       1.000000   

       total_feedback_count  total_neg_feedback_count  \
count          1.094411e+06              1.094411e+06   
mean           4.177126e+00              8.948695e-01   
std            2.271524e+01              5.288943e+00   
min            0.000000e+00              0.000000e+00   
25%            0.000000e+00              0.000000e+00   
50%            0.000000e+00              0.000000e+00   
75%            3.000000e+00              1.000000e+00   
max            5.464000e+03              1.159000e+03   

       total_pos_feedback_count  
count              1.094411e+06  
mean               3.282257e+00  
std                1.967482e+01  
min                0.000000e+00  
25%                0.000000e+00  
50%                0.000000e+00  
75%                3.000000e+00  
max                5.050000e+03  

[8 rows x 21 columns]

In [13]:
len(reviews_df)

1102905

### Columns

In [14]:
reviews_df.columns

Index(['product_id', 'product_name', 'brand_id', 'brand_name', 'loves_count',
       'rating', 'reviews', 'size', 'variation_type', 'variation_value',
       'variation_desc', 'ingredients', 'price_usd', 'value_price_usd',
       'sale_price_usd', 'limited_edition', 'new', 'online_only',
       'out_of_stock', 'sephora_exclusive', 'highlights', 'primary_category',
       'secondary_category', 'tertiary_category', 'child_count',
       'child_max_price', 'child_min_price', 'Unnamed: 0', 'author_id',
       'is_recommended', 'helpfulness', 'total_feedback_count',
       'total_neg_feedback_count', 'total_pos_feedback_count',
       'submission_time', 'review_text', 'review_title', 'skin_tone',
       'eye_color', 'skin_type', 'hair_color'],
      dtype='object')

### Count Nulls

In [15]:
print((reviews_df[:].isnull().sum() / len(reviews_df)).map('{:.2%}'.format))

percent_null = reviews_df.isnull().mean()

null_cols = percent_null[percent_null > .10].index.tolist()

print(null_cols)




product_id                   0.00%
product_name                 0.00%
brand_id                    99.23%
brand_name                   0.00%
loves_count                 99.23%
rating                       0.03%
reviews                     99.26%
size                        99.38%
variation_type              99.36%
variation_value             99.37%
variation_desc              99.89%
ingredients                 99.32%
price_usd                    0.00%
value_price_usd             99.96%
sale_price_usd              99.98%
limited_edition             99.23%
new                         99.23%
online_only                 99.23%
out_of_stock                99.23%
sephora_exclusive           99.23%
highlights                  99.43%
primary_category            99.23%
secondary_category          99.23%
tertiary_category           99.32%
child_count                 99.23%
child_max_price             99.75%
child_min_price             99.75%
Unnamed: 0                   0.77%
author_id           

# Data Cleaning

## Products

### Drop Duplicates

In [16]:
product_df.drop_duplicates(inplace=True)

#get length of products table
print(len(product_df))

8494


### Drop Columns with Large Amounts of Nulls

In [17]:
product_df.drop(columns = ['size',
                'variation_type',
                'variation_desc',
                'variation_value',
                'value_price_usd',
                'sale_price_usd',
                'highlights',
                'child_max_price',
                'child_min_price',
                'ingredients',
                'tertiary_category'
                
                
], inplace = True)



In [18]:
product_df.drop(columns = 'child_count', inplace=True)

In [19]:
product_df.drop(columns = ['reviews'], inplace = True)

In [20]:
product_df.tail()

product_id                                       product_name  brand_id  \
8489    P467659                   Couture Clutch Eyeshadow Palette      1070   
8490    P500874                              L'Homme Eau de Parfum      1070   
8491    P504428                   Mon Paris Eau de Parfum Gift Set      1070   
8492    P504448                           Y Eau de Parfum Gift Set      1070   
8493    P505461  Candy Glaze Lip Gloss Stick Duo with Hyaluroni...      1070   

              brand_name  loves_count  rating  price_usd  limited_edition  \
8489  Yves Saint Laurent         2790  4.4286      150.0                0   
8490  Yves Saint Laurent         2319  4.6367      106.0                0   
8491  Yves Saint Laurent         1475  5.0000      134.0                1   
8492  Yves Saint Laurent          840     NaN      167.0                1   
8493  Yves Saint Laurent          193     NaN       50.0                0   

      new  online_only  out_of_stock  sephora_exclusive primary_category  \
8489    0            0             0                  0           Makeup   
8490    0            0             0                  0        Fragrance   
8491    1            1             1                  0        Fragrance   
8492    1            1             0                  0        Fragrance   
8493    1            1             0                  1           Makeup   

     secondary_category  
8489                Eye  
8490                Men  
8491  Value & Gift Sets  
8492  Value & Gift Sets  
8493                Lip

### Check Products Length

In [21]:

#get length of reviews table
print(len(product_df))

8494


## Reviews

### Drop Duplicates

In [22]:
reviews_df.drop_duplicates(inplace=True)


### Drop Columns with Large Amounts of Nulls

In [23]:
reviews_df.drop(columns = null_cols, inplace=True 
               
)

In [24]:
reviews_df.drop(columns = ['Unnamed: 0', 'author_id'], inplace=True)

In [25]:
reviews_df.tail()

product_id                                   product_name  \
1102900    P500101  Hydro Ungrip Makeup Remover + Cleansing Water   
1102901    P500101  Hydro Ungrip Makeup Remover + Cleansing Water   
1102902    P500101  Hydro Ungrip Makeup Remover + Cleansing Water   
1102903    P500101  Hydro Ungrip Makeup Remover + Cleansing Water   
1102904    P500101  Hydro Ungrip Makeup Remover + Cleansing Water   

          brand_name  rating  price_usd  total_feedback_count  \
1102900  MILK MAKEUP     5.0       32.0                   0.0   
1102901  MILK MAKEUP     5.0       32.0                   0.0   
1102902  MILK MAKEUP     5.0       32.0                   0.0   
1102903  MILK MAKEUP     5.0       32.0                   0.0   
1102904  MILK MAKEUP     5.0       32.0                   0.0   

         total_neg_feedback_count  total_pos_feedback_count submission_time  \
1102900                       0.0                       0.0      2022-04-29   
1102901                       0.0                       0.0      2022-04-29   
1102902                       0.0                       0.0      2022-04-29   
1102903                       0.0                       0.0      2022-04-29   
1102904                       0.0                       0.0      2022-04-29   

                                               review_text  
1102900  I LOVE this stuff! It works amazingly well at ...  
1102901  love this micellar water from Milk! it removes...  
1102902  I loveeeeee Milk!!! Just discovered this brand...  
1102903  My new favorite makeup remover. First of all, ...  
1102904  First of all, this actually removed ALL of my ...

### Check Reviews Length

In [26]:

#get length of reviews table
print(len(reviews_df))

1102905


### Compare Length of Two Tables


In [27]:

#get length of reviews table

table_length_percent = (len(product_df) / len(reviews_df)) * 100
print(f"{table_length_percent:.2%}")

77.01%


The reviews table is noticeably longer than the products table, the products table being 70% the length of reviews. This is possible for two reasons:

1. There are > 1 reviews for each product 

2. There are reviews for products not in the products table

In [28]:
product_df['primary_category'].value_counts()

primary_category
Skincare           2420
Makeup             2369
Hair               1464
Fragrance          1432
Bath & Body         405
Mini Size           288
Men                  60
Tools & Brushes      52
Gifts                 4
Name: count, dtype: int64

In [29]:
product_df['product_id'].value_counts().loc[lambda x: x > 1]




Series([], Name: count, dtype: int64)

# Merge Tables

In [30]:

merged_df = pd.merge(reviews_df, product_df, on='product_id', how='left')



One thing I want to check is that the num_feedback columns don't have a larger amount of nulls than they say they do. This may be because NaN is being stored as a string rather than a null. To confirm, let's check:

In [31]:
merged_df.loc[:,['total_feedback_count','total_neg_feedback_count','total_pos_feedback_count']].isnull().mean()

cols = ['total_feedback_count', 'total_neg_feedback_count', 'total_pos_feedback_count']


print("Number of Nulls")
print(merged_df[cols].isnull().sum(),"\n")



print("Number of Non-Nulls")
print(merged_df[cols].notnull().sum())


Number of Nulls
total_feedback_count        8494
total_neg_feedback_count    8494
total_pos_feedback_count    8494
dtype: int64 

Number of Non-Nulls
total_feedback_count        1094411
total_neg_feedback_count    1094411
total_pos_feedback_count    1094411
dtype: int64


This confirms that only a small amount of values in these columns are null, so it may be that the nulls are highly clustered together or that the first ~8494 columns are null.

### Remove rows where review_text is null

To prepare the data for sentiment analysis, I will drop rows where review_text is null. This accounts for only about .09% of the data but will remove rows where we cannot perform the analysis.


In [32]:

merged_df['review_text'].isnull().sum()


9938

In [33]:
merged_df['primary_category'].value_counts()

primary_category
Skincare           1096831
Makeup                2369
Hair                  1464
Fragrance             1432
Bath & Body            405
Mini Size              288
Men                     60
Tools & Brushes         52
Gifts                    4
Name: count, dtype: int64

In [34]:
merged_df = merged_df.dropna(subset = ['review_text'])


In [35]:
merged_df['primary_category'].value_counts()

primary_category
Skincare    1092967
Name: count, dtype: int64

# Perform Sentiment Analysis

In [36]:
#import ntlk

from nltk.sentiment.vader import SentimentIntensityAnalyzer as SIA
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
import numpy as np
import swifter

stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

sia = SIA()

def get_sentiment(text):
    return sia.polarity_scores(text)['compound']

### Create sentiment score column

To do our analysis, I will create a new column sentiment_score, which applies sentiment analysis to each instance of the review_text column. Because this data is 1 mil+ rows, I will comment this code out and load the new merged_df into a parquet file so it can be quickly loaded in.

In [37]:
'''
merged_df['sentiment_score'] = merged_df['review_text'].astype(str).swifter.apply(
   get_sentiment
)
'''

"\nmerged_df['sentiment_score'] = merged_df['review_text'].astype(str).swifter.apply(\n   get_sentiment\n)\n"

In [38]:
merged_df

product_id                                     product_name_x  \
8494       P504322                     Gentle Hydra-Gel Face Cleanser   
8495       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
8496       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
8497       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
8498       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
...            ...                                                ...   
1102900    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102901    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102902    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102903    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102904    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   

        brand_name_x  rating_x  price_usd_x  total_feedback_count  \
8494        NUDESTIX       5.0         19.0                   2.0   
8495         LANEIGE       1.0         24.0                   0.0   
8496         LANEIGE       5.0         24.0                   0.0   
8497         LANEIGE       5.0         24.0                   0.0   
8498         LANEIGE       5.0         24.0                   0.0   
...              ...       ...          ...                   ...   
1102900  MILK MAKEUP       5.0         32.0                   0.0   
1102901  MILK MAKEUP       5.0         32.0                   0.0   
1102902  MILK MAKEUP       5.0         32.0                   0.0   
1102903  MILK MAKEUP       5.0         32.0                   0.0   
1102904  MILK MAKEUP       5.0         32.0                   0.0   

         total_neg_feedback_count  total_pos_feedback_count submission_time  \
8494                          0.0                       2.0      2023-02-01   
8495                          0.0                       0.0      2023-03-21   
8496                          0.0                       0.0      2023-03-21   
8497                          0.0                       0.0      2023-03-20   
8498                          0.0                       0.0      2023-03-20   
...                           ...                       ...             ...   
1102900                       0.0                       0.0      2022-04-29   
1102901                       0.0                       0.0      2022-04-29   
1102902                       0.0                       0.0      2022-04-29   
1102903                       0.0                       0.0      2022-04-29   
1102904                       0.0                       0.0      2022-04-29   

                                               review_text  ... loves_count  \
8494     I use this with the Nudestix “Citrus Clean Bal...  ...         177   
8495     I bought this lip mask after reading the revie...  ...     1081315   
8496     My review title says it all! I get so excited ...  ...     1081315   
8497     I’ve always loved this formula for a long time...  ...     1081315   
8498     If you have dry cracked lips, this is a must h...  ...     1081315   
...                                                    ...  ...         ...   
1102900  I LOVE this stuff! It works amazingly well at ...  ...       14632   
1102901  love this micellar water from Milk! it removes...  ...       14632   
1102902  I loveeeeee Milk!!! Just discovered this brand...  ...       14632   
1102903  My new favorite makeup remover. First of all, ...  ...       14632   
1102904  First of all, this actually removed ALL of my ...  ...       14632   

         rating_y price_usd_y  limited_edition  new  online_only  \
8494       5.0000        19.0                0    0            1   
8495       4.3508        24.0                0    0            0   
8496       4.3508        24.0                0    0            0   
8497       4.3508        24.0                0    0            0   
8498       4.3508        24.0                0    0            0

# Load merged_df into a parquet file

In [39]:

'''
merged_df.to_parquet("ulta_reviews.parquet")

'''

'\nmerged_df.to_parquet("ulta_reviews.parquet")\n\n'

### Load in merged_df from parquet

In [40]:
merged_df = pd.read_parquet("ulta_reviews.parquet")

### Clean column names

Pandas added _x and _y to the names of some of our columns, so I'm going to remove those from the column names for cleanliness

In [41]:
merged_df.columns = merged_df.columns.str.replace(r'_x$|_y$', '', regex=True)


In [42]:
merged_df

product_id                                       product_name  \
8494       P504322                     Gentle Hydra-Gel Face Cleanser   
8495       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
8496       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
8497       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
8498       P420652  Lip Sleeping Mask Intense Hydration with Vitam...   
...            ...                                                ...   
1102900    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102901    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102902    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102903    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   
1102904    P500101      Hydro Ungrip Makeup Remover + Cleansing Water   

          brand_name  rating price_usd  total_feedback_count  \
8494        NUDESTIX     5.0      19.0                   2.0   
8495         LANEIGE     1.0      24.0                   0.0   
8496         LANEIGE     5.0      24.0                   0.0   
8497         LANEIGE     5.0      24.0                   0.0   
8498         LANEIGE     5.0      24.0                   0.0   
...              ...     ...       ...                   ...   
1102900  MILK MAKEUP     5.0      32.0                   0.0   
1102901  MILK MAKEUP     5.0      32.0                   0.0   
1102902  MILK MAKEUP     5.0      32.0                   0.0   
1102903  MILK MAKEUP     5.0      32.0                   0.0   
1102904  MILK MAKEUP     5.0      32.0                   0.0   

         total_neg_feedback_count  total_pos_feedback_count submission_time  \
8494                          0.0                       2.0      2023-02-01   
8495                          0.0                       0.0      2023-03-21   
8496                          0.0                       0.0      2023-03-21   
8497                          0.0                       0.0      2023-03-20   
8498                          0.0                       0.0      2023-03-20   
...                           ...                       ...             ...   
1102900                       0.0                       0.0      2022-04-29   
1102901                       0.0                       0.0      2022-04-29   
1102902                       0.0                       0.0      2022-04-29   
1102903                       0.0                       0.0      2022-04-29   
1102904                       0.0                       0.0      2022-04-29   

                                               review_text  ... price_usd  \
8494     I use this with the Nudestix “Citrus Clean Bal...  ...      19.0   
8495     I bought this lip mask after reading the revie...  ...      24.0   
8496     My review title says it all! I get so excited ...  ...      24.0   
8497     I’ve always loved this formula for a long time...  ...      24.0   
8498     If you have dry cracked lips, this is a must h...  ...      24.0   
...                                                    ...  ...       ...   
1102900  I LOVE this stuff! It works amazingly well at ...  ...      32.0   
1102901  love this micellar water from Milk! it removes...  ...      32.0   
1102902  I loveeeeee Milk!!! Just discovered this brand...  ...      32.0   
1102903  My new favorite makeup remover. First of all, ...  ...      32.0   
1102904  First of all, this actually removed ALL of my ...  ...      32.0   

         limited_edition new  online_only  out_of_stock  sephora_exclusive  \
8494                   0   0            1             0                  0   
8495                   0   0            0             0                  1   
8496                   0   0            0             0                  1   
8497                   0   0            0             0                  1   
8498                   0   0            0             0                  1   
...                 

In [43]:
merged_df.columns

Index(['product_id', 'product_name', 'brand_name', 'rating', 'price_usd',
       'total_feedback_count', 'total_neg_feedback_count',
       'total_pos_feedback_count', 'submission_time', 'review_text',
       'product_name', 'brand_id', 'brand_name', 'loves_count', 'rating',
       'reviews', 'price_usd', 'limited_edition', 'new', 'online_only',
       'out_of_stock', 'sephora_exclusive', 'primary_category',
       'secondary_category', 'processed_text', 'sentiment_score'],
      dtype='object')

# Feature Engineering

Because each product has multiple reviews, we need to aggregate the data between reviews for each product and fine-tune our data for a model/dashboard.

In [44]:
group = merged_df.groupby('product_id')

# Safe numeric aggregation
sentiment_stats = group['sentiment_score'].agg(['mean','min', 'max'])

# Count reviews
review_counts = group['review_text'].count().rename('review_count')

# Get first price
price_usd = group['price_usd'].first()

# Get first category
first_category = group['primary_category'].first().rename('primary_category')

#get second_category

secondary_category = group['secondary_category'].first().rename('secondary_category')

# Get total feedback count

total_feedback = group['total_feedback_count'].sum()

#Get total negative feedback

total_negative_feedback = group['total_neg_feedback_count'].sum()

#Get total positive feedback

total_positive_feedback = group['total_pos_feedback_count'].sum()

# Get total loves count

total_loves_count = group['loves_count'].sum()

#get boolean columns

limited_edition = group['limited_edition'].first().rename('limited_edition')
new_flag= group['new'].first().rename('new')
online_only = group['online_only'].first().rename('online_only')
out_of_stock = group['out_of_stock'].first().rename('out_of_stock')



In [45]:
#concat

product_scores = pd.concat([sentiment_stats, review_counts, price_usd, first_category, secondary_category, total_feedback,total_negative_feedback, total_positive_feedback,
                            total_loves_count, limited_edition, new_flag, online_only, out_of_stock], axis=1)

# Extract one row per product from merged_df
product_info = merged_df.drop_duplicates('product_id')[['product_id', 'product_name','brand_name']]

# Merge it back
product_scores = product_scores.reset_index().merge(product_info, on='product_id', how='left')

In [46]:
product_scores.rename(columns = {'mean':'mean_sentiment_score', 'max':'max_sentiment_score', 'min':'min_sentient_score'}, inplace=True)

In [47]:
product_scores = product_scores.loc[:, ~product_scores.columns.duplicated()]


In [48]:
product_scores.groupby('primary_category')['mean_sentiment_score'].median()
#product_scores.groupby('secondary_category')['mean_sentiment_score'].median()


primary_category
Skincare    0.678307
Name: mean_sentiment_score, dtype: float64

In [49]:
product_scores

product_id  mean_sentiment_score  min_sentient_score  \
0       P107306              0.511119             -0.9042   
1       P114902              0.588713             -0.9732   
2        P12045              0.649191             -0.9875   
3       P122651              0.528114             -0.8685   
4       P122661              0.608024             -0.9359   
...         ...                   ...                 ...   
2346     P94421              0.656337             -0.9628   
2347     P94812              0.739646             -0.8741   
2348      P9939              0.614457             -0.9284   
2349      P9940              0.685338             -0.9324   
2350      P9941              0.663296             -0.7851   

      max_sentiment_score  review_count  price_usd primary_category  \
0                  0.9925           253       89.0         Skincare   
1                  0.9981          1529       48.0         Skincare   
2                  0.9975          1684       12.0         Skincare   
3                  0.9829           200       20.0         Skincare   
4                  0.9895           810       26.0         Skincare   
...                   ...           ...        ...              ...   
2346               0.9973          2912       82.0         Skincare   
2347               0.9944           523       39.0         Skincare   
2348               0.9944           272       42.0         Skincare   
2349               0.9974           828       68.0         Skincare   
2350               0.9957           391       44.0         Skincare   

     secondary_category  total_feedback_count  total_neg_feedback_count  \
0              Eye Care                1906.0                     374.0   
1            Treatments                4202.0                     629.0   
2          Moisturizers                5259.0                    1050.0   
3             Cleansers                 637.0                     160.0   
4             Cleansers                1155.0                     318.0   
...                 ...                   ...                       ...   
2346         Treatments               15760.0                    3074.0   
2347          Cleansers                1252.0                     299.0   
2348          Cleansers                 841.0                     207.0   
2349       Moisturizers                1903.0                     326.0   
2350          Cleansers                1175.0                     215.0   

      total_pos_feedback_count  loves_count  limited_edition  new  \
0                       1532.0      3342636                0    0   
1                       3573.0     87084195                0    0   
2                       4209.0    235800416                0    0   
3                        477.0      1513800                0    0   
4                        837.0     13491360                0    0   
...                        ...          ...              ...  ...   
2346                   12686.0    484623776                0    0   
2347                     953.0     13103242                0    0   
2348                     634.0      4121888                0    0   
2349                    1577.0     20343960                0    0   
2350                     960.0      7134186                0    0   

      online_only  out_of_stock  \
0               0             0   
1               0             0   
2               0             0   
3               0             0   
4               0             0   
...           ...           ...   
2346            0             0   
2347            0             0   
2348            0             0   
2349            0             0   
2350            0             0   

                                           product_name         brand_name  
0                                    Renewing Eye Cream              Murad  
1          Goodbye Acne  Max Complexion Correction Pads  Peter Thomas Roth  
2                    Grape Water Mois

# Load optimized df to csv

In [50]:
#product_scores.to_csv("product_scores_input.csv", index=False)
